squirrel/sqast.h

```diff
@@ -0,0 +1,1072 @@
+        case TO_PARAM:
+            visitor->visitParamDecl(static_cast<ParamDecl *>(this)); return;
+        case TO_CONST:
+            visitor->visitConstDecl(static_cast<ConstDecl *>(this)); return;
+        case TO_DECL_GROUP:
+            visitor->visitDeclGroup(static_cast<DeclGroup *>(this)); return;
+        case TO_DESTRUCT:
+            visitor->visitDesctructionDecl(static_cast<DesctructionDecl *>(this)); return;
+        case TO_FUNCTION:
+            visitor->visitFunctionDecl(static_cast<FunctionDecl *>(this)); return;
+        case TO_CONSTRUCTOR:
+            visitor->visitConstructorDecl(static_cast<ConstructorDecl *>(this)); return;
+        case TO_CLASS:
+            visitor->visitClassDecl(static_cast<ClassDecl *>(this)); return;
+        case TO_ENUM:
+            visitor->visitEnumDecl(static_cast<EnumDecl *>(this)); return;
+        case TO_TABLE:
+            visitor->visitTableDecl(static_cast<TableDecl *>(this)); return;
+        default:
+            break;
+        }
+    }
+
+    void visitChildren(Visitor *visitor);
+
+    bool isDeclaration() const { return _op > TO_EXPR_MARK; }
+    bool isStatement() const { return _op < TO_STATEMENT_MARK; }
+    bool isExpression() const { return TO_STATEMENT_MARK < _op && _op < TO_EXPR_MARK; }
+
+    Id *asId() { assert(_op == TO_ID); return (Id*)this; }
+    GetFieldExpr *asGetField() { assert(_op == TO_GETFIELD); return (GetFieldExpr*)this; }
+    GetTableExpr *asGetTable() { assert(_op == TO_GETTABLE); return (GetTableExpr*)this; }
+
+    SQInteger linePos() const { return _linepos; }
+    void setLinePos(SQInteger pos) { _linepos = pos; }
+
+private:
+
+    SQInteger _linepos;
+
+    enum TreeOp _op;
+};
+
+class AccessExpr;
+
+class Expr : public Node {
+protected:
+    Expr(enum TreeOp op) : Node(op) {}
+
+public:
+    bool isAccessExpr() const { return TO_GETFIELD <= op() && op() <= TO_SETTABLE; }
+    AccessExpr *asAccessExpr() const { assert(isAccessExpr()); return (AccessExpr*)this; }
+};
+
+enum IdType : SQInteger {
+    ID_LOCAL = -1,
+    ID_CONST = -2,
+    ID_FIELD = -3
+};
+
+class Id : public Expr {
+public:
+    Id(const SQChar *id) : Expr(TO_ID), _id(id), _outpos(ID_LOCAL), _assignable(false) {}
+
+    void visitChildren(Visitor *visitor) {}
+
+    const SQChar *id() { return _id; }
+
+    void setOuterPos(SQInteger pos) { _outpos = pos; }
+    
+    bool isOuter() const { return _outpos >= 0; }
+    bool isField() const { return _outpos == ID_FIELD; }
+    void setField() { _outpos = ID_FIELD; }
+    bool isLocal() const { return _outpos == ID_LOCAL; }
+    bool isConst() const { return _outpos == ID_CONST; }
+    void setConst() { _outpos = ID_CONST; }
+
+    SQInteger outerPos() const { return _outpos; }
+    void setAssiagnable(bool v) { _assignable = v; }
+    bool isAssignable() const { return _assignable; }
+    bool isBinding() const { return (isOuter() || isLocal()) && isAssignable(); }
```

Comment:
Should be `!isAssignable()`
Assignability is what differs variables (declared with `local`) from named bindings (declared with `let`)

Reply:
fixed